base/shell/explorer/trayntfy.cpp

```diff
@@ -308,17 +313,226 @@ class CIconWatcher
     }
 };
 
+class CBalloonQueue
+{
+public:
+    static const int TimerInterval = 2000;
+    static const int BalloonsTimerId = 1;
+    static const int MinTimeout = 10000;
+    static const int MaxTimeout = 30000;
+    static const int CooldownBetweenBalloons = 2000;
+
+private:
+    struct Info
+    {
+        InternalIconData * pSource;
+        WCHAR szInfo[256];
+        WCHAR szInfoTitle[64];
+        WPARAM uIcon;
+        UINT uTimeout;
+
+        Info(InternalIconData * source)
+        {
+            pSource = source;
+            StrNCpy(szInfo, source->szInfo, _countof(szInfo));
+            StrNCpy(szInfoTitle, source->szInfoTitle, _countof(szInfoTitle));
+            uIcon = source->dwInfoFlags & 3;
+            if (source->dwInfoFlags & NIIF_USER)
+                uIcon = reinterpret_cast<WPARAM>(source->hIcon);
+            uTimeout = source->uTimeout;
+        }
+    };
+
+    HWND m_hwndParent;
+
+    CTooltips * m_tooltips;
+
+    CAtlList<Info> m_queue;
+
+    CToolbar<InternalIconData> * m_toolbar;
+
+    InternalIconData * m_current;
+    bool m_closed;
+
+    int m_timer;
+
+public:
+    CBalloonQueue() :
+        m_hwndParent(NULL),
+        m_tooltips(NULL),
+        m_toolbar(NULL),
+        m_current(NULL),
+        m_closed(false),
+        m_timer(-1)
+    {
+    }
+
+    void Init(HWND hwndParent, CToolbar<InternalIconData> * toolbar, CTooltips * balloons)
+    {
+        m_hwndParent = hwndParent;
+        m_toolbar = toolbar;
+        m_tooltips = balloons;
+    }
+
+    void Deinit()
+    {
+        if (m_timer >= 0)
+        {
+            ::KillTimer(m_hwndParent, m_timer);
+        }
+    }
+
+    bool OnTimer(int timerId)
+    {
+        if (timerId != m_timer)
+            return false;
+
+        ::KillTimer(m_hwndParent, m_timer);
+        m_timer = -1;
+
+        if (m_current && !m_closed)
+        {
+            Close(m_current);
+        }
+        else
+        {
+            m_current = NULL;
+            m_closed = false;
+            if (!m_queue.IsEmpty())
+            {
+                Info info = m_queue.RemoveHead();
+                Show(info);
+            }
+        }
+
+        return true;
+    }
+
+    void UpdateInfo(InternalIconData* notifyItem)
+    {
+        size_t len = 0;
+        bool hasInfo = SUCCEEDED(StringCchLength(notifyItem->szInfo, _countof(notifyItem->szInfo), &len)) && len > 0;
+        if (m_current == notifyItem)
+        {
+            if (hasInfo)
+            {
+                Info info(notifyItem);
+                Show(info);
+            }
+            else
+            {
+                Close(notifyItem);
+            }
+        }
+        else
+        {
+            if (hasInfo)
+            {
+                if (m_current || !m_queue.IsEmpty())
+                {
+                    m_queue.AddTail(Info(notifyItem));
```

Comment:
You use a return value here in contrast to temporary variables in other places.

Reply:
That was because the other places forced me to use an intermediate variable, but this one didn't. But regardless, it's unified now, so the situation has been removed.